src/db-api.js

```diff
@@ -64,62 +91,165 @@ const transactionsHistoryForAddresses = (db: Pool) => async (
+    values: [tx],
+  })
+
+/**
+* Queries TX_BODY table looking for tx body by its hash
+* @param {Db Object} db
+* @param {Transaction} tx
+*/
+const getRawTx = (db: Pool) => async (tx: string): Promise<ResultSet> =>
+  db.query({
+    text: 'SELECT body::text as tx_body FROM tx_body WHERE hash = $1',
+    values: [tx],
+  })
+
+/**
+* Queries BLOCK table looking for block with a given id
+* @param {Db Object} db
+* @param {Block} blockId
+*/
+const getBlockById = (db: Pool) => async (blockId: number): Promise<ResultSet> =>
+  db.query({
+    text: 'SELECT time, block_no, hash::text FROM block WHERE id = $1',
+    values: [blockId],
+  })
+
+/**
+* Quries TX* tables to get txInputs for a given transaction
+* @param {Db Object} db
+* @param {Transaction} tx
+*/
+const getSingleTxInputs = (db: Pool) => async (tx: number): Promise<ResultSet> =>
+  db.query({
+    text: `SELECT
+      tx_out.address, tx_out.value
+      FROM tx_out
+      INNER JOIN tx ON tx.id = tx_out.tx_id
+      INNER JOIN tx_in ON tx_in.tx_out_id = tx_out.tx_id AND tx_in.tx_out_index = tx_out.index
+      WHERE tx_in.tx_in_id = $1`,
+    values: [tx],
+  })
+
+/**
+* Queries TX, BLOCK, TX_OUT tables to acquire inward and outward transactions for given addresses
+* @param {Db Object} db
+* @param {Array<Address>} addresses
+*/
+const getTransactions = (db: Pool) => async (addresses: Array<string>): Promise<ResultSet> =>
+  db.query({
+    text: `SELECT
+      tx.id, tx.hash::text, block.time
+      FROM block 
+      INNER JOIN tx ON block.id = tx.block 
+      INNER JOIN tx_out ON tx.id = tx_out.tx_id
+      WHERE tx_out.address = ANY($1)
+    UNION
+    SELECT DISTINCT 
+      tx.id, tx.hash::text, block.time
+      FROM block 
+      INNER JOIN tx ON block.id = tx.block 
+      INNER JOIN tx_in ON tx.id = tx_in.tx_in_id 
+      INNER JOIN tx_out ON (tx_in.tx_out_id = tx_out.tx_id) AND (tx_in.tx_out_index = tx_out.index)
+      WHERE tx_out.address = ANY($1)`,
+    values: [addresses],
+  })
+
+/**
+* Queries TX* tables to acquire bulk tx inputs for given transactions
+* @param {Db Object} db
+* @param {Array<Transaction>} txIds
+*/
+const getTxsInputs = (db: Pool) => async (txIds: Array<number>): Promise<ResultSet> =>
+  db.query({
+    text: `SELECT DISTINCT
+      tx.id as txId, tx_out.address, tx_out.value, tx2.hash::text, tx_out.index, (tx2.size = 0) 
+      FROM tx
+      INNER JOIN tx_in ON tx.id = tx_in.tx_in_id 
+      INNER JOIN tx_out ON (tx_in.tx_out_id = tx_out.tx_id) AND (tx_in.tx_out_index = tx_out.index) 
+      INNER JOIN tx AS tx2 ON tx2.id = tx_in.tx_out_id
+      WHERE tx_in.tx_in_id = ANY($1)`,
+    values: [txIds],
+  })
+
+/**
+* Queries TX* tables to acquire bulk tx outputs for given transactions
+* @param {Db Object} db
+* @param {Array<Transaction>} txIds
+*/
+const getTxsOutputs = (db: Pool) => async (txIds: Array<number>): Promise<ResultSet> =>
+  db.query({
+    text: `SELECT
+      tx.id as txId, tx_out.address, tx_out.value
```

Comment:
I would select also the index to order by it in the response- it's kind of weird to return the outputs out of order

Reply:
bear in mind that ordering in the SQL result is not useful as the results are grouped by tx id on application level, so the ordering by index should probably happen there as well: https://stackoverflow.com/questions/28860526/lodash-sortby-then-groupby-is-order-maintained

Same holds for returning the txsInputs in order